018-cli-2.0/018-cli-2.0.md

```diff
@@ -0,0 +1,393 @@
+
+The refactored command will look like this 
+
+```sh
+appwrite functions createTag --functionId=[ID] --command=[COMMAND]
+```
+
+This corresponding appwrite issue will also need to be tackled as we implement this.
+https://github.com/appwrite/appwrite/issues/1316
+
+### 🟢 The `appwrite.json` File
+
+The next improvement is going to be along the lines of a `appwrite.json` file. Running `appwrite init` in the current directory will create a `appwrite.json` file and initialize the current directory with attributes necessary to communicate with a particular project.
+
+Consider this example
+```sh
+mkdir temp && cd temp
+
+# Initialize the current directory with an Appwrite project 
+appwrite init
+
+temp 
+└── appwrite.json
+```
+
+A `appwrite.json` file will be associated with one project only whose ID will be visible in the `project` attribute. 
+
+```json
+{
+    "project" : "",
+    "functions": [
+        {
+          "name" : "My awesome function 💪",
+          "path" : "./FunctionOne",
+          "command" : "python main.py",
+          "runtime": "python3.9",
+          // This section is useful when we allow cloud functions to be tested locally
+          "vars" :{
+            "KEY 1" : "VALUE 1",
+          }
+        },
+    ],
+    "hosting": []
+}
+```
+
+There are two scenarios here 
+* A developer would like to link the directory to an existing project
+
+```sh
+appwrite init
+
+Choose a project to link this directory to 
+⭕️ Project 1
+✅️ Project 2
+⭕️ Project 3
+```
+
+* A developer would like to create a new project while initializing this directory
+
+```sh
+appwrite init --new
+
+Give your project a name:
+Project X
+
+Creating Project X...
+
+Project X Created 👍
+```
+
+> We need to first add support to utopia-php/CLI to allow for single select and multi select commands to enable this feature.
+
+
+### 🟢 The `appwrite deploy` command
+
+The `appwrite deploy` command is a convenient wrapper around three main commands
+* Create Function
+* Create Tag
+* Update Tag
+
+```sh
+# Usage 1
+appwrite deploy functions
+Which functions would you like to deploy? 
+✅️ FunctionOne
+✅️ FunctionTwo
+⭕️ FunctionThree
+
+# Usage 2
+appwrite deploy functions --all
+# Deploys all the functions specified in appwrite.json
+
+# Future support for static hosting
+appwrite deploy site
+appwrite deploy site --all 
+```
+
+## Easier creation of cloud functions
+---
+
+We will introduce an `appwrite generate` command that does the following
+* Generate a function / static site template. 
+* Make an entry in the `appwrite.json` file. Throw an error if the file doesn't exist.
+
+Here are some example usages 
+```sh
+# Usage 1
+appwrite generate function --name="FunctionOne" --runtime=python-3.9
+# Creates a folder called FunctionOne with a template for a sample python function.
+temp
+├── appwrite.json
+└── FunctionOne
+    ├── .appwrite
```

Comment:
May be some information regarding `.appwrite` file will be useful as well

Reply:
Is it possible to have everything in `appwrite.json`? I think it's a bit confusing to have two different types of config files.

Reply:
Maybe most major configs are done using appwrite.json then .appwrite can be used for API keys and project keys as it'll be hidden by default right? Aslong as we remind users not to commit .appwrite they can share their project with ease

Reply:
Agreeing with @kodumbeats - If we are going to have a `.appwrite` directory, why not have `config.yaml` in there (example, how .git is setup) or just have `appwrite.yaml`

Reply:
@lohanidamodar  `.appwrite` is actually the folder which some runtimes use to store their dependencies like Python and ruby  . Similar to node modules in npm projects

Reply:
@kodumbeats `appwrite.json` is a file that is expected to be a part of version control since it will have all the information related to a project like the collections, functions and so on and hence allow them to recreate their project environment . Having secrets in this file doesn't seem like a good idea